support nesting of suppress_guards, suppress guards when generated compiled autograd graph

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/138920. See comments there for details.

I still need to try to get a smaller repro to write an actual test. But suppressing the guards, I now no longer see the specilization in the CA graph in the linked example:
```
        aot1_view_3: ... = torch.ops.aten.view.default(aot1_tangents_1, [aot1_sym_size_int, 48, 1])
        aot1_view_4: ... = torch.ops.aten.view.default(aot1_view_3, [aot1_sym_size_int, 48])
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138968



cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @xmfan @rec

bdhirsh committed:
support nesting of suppress_guards, suppress guards when generated compiled autograd graph

[ghstack-poisoned]

yf225 commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -1,6 +1,7 @@
 # mypy: allow-untyped-defs
```
I also feel that the repro in https://github.com/pytorch/pytorch/issues/138920 isn't too long, and maybe we can add it as a unit test in this PR

bdhirsh commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -227,6 +228,9 @@ def begin_capture(
         self.stack.enter_context(disable_autocast_cache())
+        # Needed to make sure we don't accidentally specialize any symbols
+        assert self.fake_tensor_mode.shape_env is not None
+        self.fake_tensor_mode.shape_env._suppress_guards_enter()
```
I originally tried that and got this:
```
  File "/home/hirsheybar/local/c/pytorch/torch/_dynamo/compiled_autograd.py", line 235, in begin_capture
    self.stack.enter_context(self.fake_tensor_mode.shape_env.suppress_guards)
  File "/home/hirsheybar/.conda/envs/pytorch_c/lib/python3.10/contextlib.py", line 491, in enter_context
    _exit = _cm_type.__exit__
AttributeError: type object 'method' has no attribute '__exit__'. Did you mean: '__init__'?
```

It looks like `enter_context(x)` requires the arg that you pass into to have an `__enter__` and `__exit__`, and... it looks like wrapping a python class method with `@contextmanager` doesn't actually give it those methods.

I'm open to ideas :)

xmfan replied:
a layer of indirection works for me

```python
@contextlib.contextmanager
def annotated_fn():
    try:
        print("annotated_fn enter")
        yield
    finally:
        print("annotated_fn close")

class MyClass():
    def get_annotated_fn(self):
        return annotated_fn()
```

I can probably apply this more broadly to the code base, sounds like @contextmanager methods aren't as supported

yf225 commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -374,6 +378,27 @@ def move_graph_nodes_to_cuda(self, graph) -> List[int]:
+            if (
+                node.op == "call_function"
+                and node.target == operator.eq
+                and self.is_sym_node(node.args[1])
```
Shuai found out that the sym node can be on LHS as well, we might need:
![image](https://github.com/user-attachments/assets/61ee7a6f-ae2d-41de-98f1-f2d5d755829a)

bdhirsh committed:
Update on "support nesting of suppress_guards, suppress guards when generated compiled autograd graph"

Fixes https://github.com/pytorch/pytorch/issues/138920. See comments there for details.

I still need to try to get a smaller repro to write an actual test. But suppressing the guards, I now no longer see the specilization in the CA graph in the linked example:
```
        aot1_view_3: ... = torch.ops.aten.view.default(aot1_tangents_1, [aot1_sym_size_int, 48, 1])
        aot1_view_4: ... = torch.ops.aten.view.default(aot1_view_3, [aot1_sym_size_int, 48])
```




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames rec xmfan yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "support nesting of suppress_guards, suppress guards when generated compiled autograd graph"

Fixes https://github.com/pytorch/pytorch/issues/138920. See comments there for details.

I still need to try to get a smaller repro to write an actual test. But suppressing the guards, I now no longer see the specilization in the CA graph in the linked example:
```
        aot1_view_3: ... = torch.ops.aten.view.default(aot1_tangents_1, [aot1_sym_size_int, 48, 1])
        aot1_view_4: ... = torch.ops.aten.view.default(aot1_view_3, [aot1_sym_size_int, 48])
```




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov xmfan rec

[ghstack-poisoned]

bdhirsh committed:
Update on "support nesting of suppress_guards, suppress guards when generated compiled autograd graph"

Fixes https://github.com/pytorch/pytorch/issues/138920. See comments there for details.

I still need to try to get a smaller repro to write an actual test. But suppressing the guards, I now no longer see the specilization in the CA graph in the linked example:
```
        aot1_view_3: ... = torch.ops.aten.view.default(aot1_tangents_1, [aot1_sym_size_int, 48, 1])
        aot1_view_4: ... = torch.ops.aten.view.default(aot1_view_3, [aot1_sym_size_int, 48])
```




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov xmfan rec

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge -f "inductor_torchbench_smoketest_perf CI job doesn't run the compiled autograd code path"
